File: ext/flocor/layout2d.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2021 all rights reserved


// externals
#include "external.h"
// namespace setup
#include "forward.h"
// libflocor
#include <flocor/products.h>


// type aliases
namespace flocor::py {
    // 2d layouts
    using layout2d_t = products::layout_t<2>;
    // the corresponding shape
    using shape2d_t = layout2d_t::shape_type;
    // and the index
    using index2d_t = layout2d_t::index_type;
}


// add bindings for the grid layouts used in this package
void
flocor::py::layout2d(py::module & m)
{
    // 2d layouts
    auto layout = py::class_<layout2d_t>(m, "Layout2D");

    // constructors
    // from a shape
    layout.def(
        // the handler
        py::init<const shape2d_t &>(),
        // the signature
        "shape"_a);

    // from a shape and an origin
    layout.def(
        // the handler
        py::init<const shape2d_t &, const index2d_t &>(),
        // the signature
        "shape"_a, "origin"_a);

    // accessors
    // my shape
    layout.def_property_readonly(
        // the name
        "shape",
        // the getter
        &layout2d_t::shape,
        // the docstring
        "get my shape");

    // my origin
    layout.def_property_readonly(
        // the name
        "origin",
        // the getter
        &layout2d_t::origin,
        // the docstring
        "get my origin");

    // all done
    return;
}


// end of file

File: tests/flocor.ext/layout2d.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2021 all rights reserved


# the driver
def test():
    """
    Verify we can build and use 2D layouts
    """
    # get the package
    from flocor.ext import libflocor

    # verify that the layout object bindings exist
    assert libflocor.Layout2D

    # pick a shape
    dim = 256,128
    dim = 2,3
    # and the layout origin
    org = -2, 17

    # make a shape
    shape = libflocor.Shape2D(shape=dim)
    # and an origin
    origin = libflocor.Index2D(index=org)
    # use it to make a layout
    layout = libflocor.Layout2D(shape=shape, origin=origin)

    # check the shape
    assert tuple(layout.shape) == dim
    # the origin
    assert tuple(layout.origin) == org
    # the strides
    assert tuple(layout.strides) == (dim[1], 1)
    # and the nudge, using its definition
    assert layout.nudge == layout.offset(index=(0,0))

    # iterate over the layout in its native order
    for index in layout:
        # verify the index identity; convert to tuples since we don't have {__eq__} yet
        assert tuple(layout.index(layout.offset(index))) == tuple(index)

    # go through all possible offsets
    for offset in range(layout.shape.cells):
        # verify the offset identity
        assert layout.offset(layout.index(offset)) == offset

    # all done
    return 0


# bootstrap
if __name__ == "__main__":
    # invoke the driver
    status = test()
    # and share the status with the shell
    raise SystemExit(status)


# end of file
